Add optional litespi wrapping into separate clock domain

kowalewskijan opened the pull request:
This PR introduces changes which give a possibility to have LiteSPI in a separate clock domain.
The only doubt I have is do we have to pipeline the data from CSRs (which are in sys clock domain) to litespi clock domain using e.g. AsynFIFOs? I mean these [registers](https://github.com/litex-hub/litespi/pull/38/files#diff-76745ba1ce398bffc095fd8450e49fcdL51-L52). I tested on Arty board two configurations:
- LiteSPI in `sys` clock domain
- LiteSPI in `clk200` clock domain

in each cases I was able to successfully read from flash.

Closes #23

kowalewskijan commented:
I made some major changes in code:
- Reworked `LiteSPICore` to use it to instantiate Master/MMAP cores and to handle crossbar connections there ([link](https://github.com/litex-hub/litespi/pull/38/files#diff-76745ba1ce398bffc095fd8450e49fcdR13-R82))
- Reworked `LiteSPI` wrapper - now its primary application is to handle optional litespi domain and to instantiate `LiteSPICore` and add it along with PHY to submodules ([link](https://github.com/litex-hub/litespi/pull/38/files#diff-76745ba1ce398bffc095fd8450e49fcdR116-R134))
- Added MultiReg support for cross-domain R/Ws to PHY CSRs ([link](https://github.com/litex-hub/litespi/pull/38/files#diff-623560220a63e762721c7198f091b071R90-R111))
- Added AsyncFIFOs in crossbar to handle cross-domain streams ([link](https://github.com/litex-hub/litespi/pull/38/files#diff-6f3ec1c373bdeb333e77783e96035796R33-R42))
- Updated docstrings

enjoy-digital commented:
This looks good to me. We generally declare the `submodules` in the `Module` they are defined, here it is passed to `LiteSPI` and declared there. I think you are doing this to do the clock domain renaming in `LiteSPI`, but you could probably just define the `litespi` clock domain in the PHY, use this in it and re-reclare the PHY as a `submodule` in the target. If this make senses for you, can you just change that?

kowalewskijan commented:
- I extracted logic part of PHY into separate module so I could use renamer inside the PHY ([link](https://github.com/litex-hub/litespi/pull/38/files#diff-623560220a63e762721c7198f091b071R26-R275)).
- Now in `LiteSPIPHY` we just instantiate logic part of PHY as module and we can handle optional clock domain ([link](https://github.com/litex-hub/litespi/pull/38/files#diff-623560220a63e762721c7198f091b071R313-R329))
- I moved around adding to submodules, so we have logic part of PHY added to submodule in PHY part like [here](https://github.com/litex-hub/litespi/pull/38/files#diff-623560220a63e762721c7198f091b071R326) and we have a redeclaration of PHY submodule in target like [here](https://github.com/litex-hub/litespi/pull/38/files#diff-de492c7b09d278e7cf3335d6b1f192c0R93)

piotr-binkowski commented on litespi/__init__.py:
```diff
@@ -71,10 +70,57 @@ def __init__(self, phy, sys_clk_freq, with_mmap=True, with_master=True, mmap_end
+        self.bus = self.core.bus
+
+        self._csrs = []
+        self._csrs = self.core.get_csrs() + phy.get_csrs()
```
I think this `LiteSPI` wrapper is no longer needed since the PHY now can handle clock domains on its own

kowalewskijan replied:
I reverted to original state. So we have just `LiteSPI`. CSRs for PHY and cores are separated [link](https://github.com/litex-hub/litespi/pull/38/files#diff-de492c7b09d278e7cf3335d6b1f192c0R89-R94)

kowalewskijan commented:
CSRs for PHY and cores are separated but this results in different function names for PHY CSRs. I will prepare a PR to LiteX to fix this issue.

enjoy-digital commented:
@kowalewskijan: thanks for the update, since i think @piotr-binkowski is also ok with it, let's merge.
